internal/expressions/json_templates.go

```diff
@@ -155,15 +155,20 @@ func EvaluateTemplate(template string, env map[string]any, exprOpts ...expr.Opti
 		}
 		return result, nil
 	}
-	// If the result is parseable as a bool return that.
-	if resBool, err := strconv.ParseBool(result); err == nil {
-		return resBool, nil
-	}
+
+	// NOTE: The result is explicitly parsed as numbers before booleans to prevent values like
+	// "0" and "1" from being misinterpreted as boolean `false` and `true`,
+	// respectively. See: https://github.com/akuity/kargo/pull/3516
+
 	// If the result is parseable as a float64, return that. float64 is used
 	// because it can represent all JSON numbers.
```

Comment:
```suggestion
	// If the result is parseable as a float64, return that. float64 is used
	// because it can represent all JSON numbers.
	//
	// NB: This is attempted prior to attempting to parse the result as a boolean
	// so that "0" and "1" will be interpreted as numbers.
```

Reply:
I don't think we need a reference to the PR living in here forever when the comment explains everything there is to know about why we're doing this quite clearly.